Design patient registry

lomamech opened the pull request:
- Adding unit patient search and showing the data in a UI GRID
- Adding test E2E for research patient
- Update some test E2E

@jniles

lomamech committed:
Adding unit patient search and showing the data in a UI GRID

lomamech committed:
Adding test E2E for research patient

lomamech committed:
Improving the test E2E for Patient Registry and Adding an ID on the UI GRID for Patients Registries

lomamech committed:
Update some test E2E
Update the condition of comparaison for billing service and debtors groups
update the rank for the deletion for the unit subsidy

jniles commented on client/src/i18n/fr.json:
```diff
@@ -536,6 +536,15 @@
+    "MONTHS"                : "mois",
+    "YEAR_OLD"              : "an",
+    "YEARS_OLD"             : "ans",
+    "PAGE_TITLE"            : "Recherche des patients"
```
Normalement, nous appelons cette clé "TITLE".

jniles commented on client/src/js/app.js:
```diff
@@ -538,8 +538,9 @@ function bhimaConfig($stateProvider, $urlRouterProvider, $urlMatcherFactoryProvi
 
-  .state('/patients/search/:patientID?', {
-    controller: 'patientRecords',
+  .state('/patients/search/', {
```
Ce PR correspond a l'issue #336 qui décrit une module qui s'appele "Patient Registry".  En Francais, ca signifie "Registre des Patients".  Le meilleur url est `/patients/registry` avec un nom `PatientRegistryController`.

jniles commented on client/src/js/services/DateService.js:
```diff
@@ -45,6 +45,7 @@ function DateService() {
   service.current = {};
   service.next = {};
   service.util = {};
+  service.getAge = getAge;
```
Je pense cette fonctionnalité a été supplimmer parce que `angular-moment` peut fourni la meme fonctionnalité.  Essayer en peut pour voir si `angular-moment` peut le fair.

sfount replied:
I think I have come across this before! 

I found this reference very useful http://momentjs.com/docs/#/displaying/difference/

As I understand it you can perform a query like: 

``` js
/**
* returns the number of years between now (specified by moment() just like Date())
* and the specified date passed as the first parameter to the diff method
*/
moment().diff( date, 'years' );
```

It is worth noting that because we include the Angular Moment library the full moment library can be required by a controller/ service by injecting `moment` 

``` js
Controller.$inject = ['moment'];

function Controller(moment) { 
  // ...
}
```

Let me know if there's anything that doesn't make sense!

jniles replied:
Thanks for the documentation reference @sfount!

jniles commented on client/src/js/services/DateService.js:
```diff
@@ -217,7 +218,7 @@ function DateService() {
   // expects a date object
   // return 'YYYY-MM-DD' format
-  service.util.str = function (date) {
+  service.util.str = function (date, simple) {
```
Pourquoi avez-vous ajouté la parametre `simple`?

jniles commented on client/src/js/services/PatientService.js:
```diff
@@ -47,7 +47,7 @@ function PatientService($http, util, Session) {
    */
   function detail(uuid) {
-    return $http.get(baseUrl.concat(uuid))
+    return $http.get(baseUrl.concat(uuid || ''))
```
Good catch!

jniles commented on client/src/partials/patients/search/search.js:
```diff
@@ -0,0 +1,65 @@
+];
+
+/** 
+ * Users and Permission Controller 
```
Ce module n'est pas le "Users and Permissions Controller".  S'il vous plait, soyez vigilant quand vous copiez et collez code.

jniles commented on client/src/partials/patients/search/search.js:
```diff
@@ -0,0 +1,65 @@
+  };
+
+  /** @todo manage state without strings */
+  vm.state = 'default'; // this is default || create || update
```
Je pense que `vm.state` n'est jamais utilisé.  Pouvez-vous l'enlever?

jniles commented on client/src/partials/patients/search/search.js:
```diff
@@ -0,0 +1,65 @@
+
+
+  // TODO
+  function handler(error) {
```
Je pense que `handler()` n'est jamais utilisé.  Pouvez-vous l'utiliser?

jniles commented on client/src/partials/patients/search/search.js:
```diff
@@ -0,0 +1,65 @@
+    throw error;
+  }
+
+  // load user grid
```
Le grid n'est pas the "user grid".  Enlevez ou mettez les commentaires a jour.

jniles commented on client/src/partials/patients/search/search.js:
```diff
@@ -0,0 +1,65 @@
+
+  // load user grid
+  function loadGrid() {
+    Patients.detail().then(function (patients) {
```
Est-ce qu'il y a une méthode `Patients.read()`?  Je pense que cette méthode est plus semantique.  Si il n'existe pas, ne hésitez pas à le créer.

jniles commented on client/test/e2e/patient/search.spec.js:
```diff
@@ -0,0 +1,21 @@
+
+helpers.configure(chai);
+
+describe('Research Patient UI Grid ', function () {
```
"Patient Registry" peut-etre une meilleure commentaire.

jniles commented on server/controllers/medical/patient.js:
```diff
@@ -327,9 +327,14 @@ function list(req, res, next) {
-    `SELECT p.uuid, CONCAT(pr.abbr, p.reference) AS patientRef, p.first_name,
-      p.middle_name, p.last_name
-    FROM patient AS p JOIN project AS pr ON p.project_id = pr.id`;
+    `SELECT p.reference, CONCAT(p.first_name,' ', p.last_name,' ', p.middle_name) AS patientName, p.uuid,
```
`p.uuid` doit être `BUID(p.uuid) AS uuid`.

lomamech committed:
Improve patient registry

Update translate files
Using patient registry instead patient search
Remove method getDate instead the using of angular-moment
Fix some comment

lomamech committed:
Complete translate liste

jniles commented on client/src/js/app.js:
```diff
@@ -538,9 +538,10 @@ function bhimaConfig($stateProvider, $urlRouterProvider, $urlMatcherFactoryProvi
-  .state('/patients/search/:patientID?', {
-    controller: 'patientRecords',
-    templateUrl: '/partials/patients/search/search.html'
+  .state('/patients/registry/', {
```
Parfait!
